bin/parallel.py

```diff
@@ -121,30 +121,24 @@ def _interrupt_handler(self, sig, frame):
 
     # Add one task. Higher priority => done first
     def put(self, task, priority=0):
+        # no task should be added after .done() was called
+        assert not self.finish
```

Comment:
This is now outside of the mutex in the parallel case, which could lead to race conditions?
(same with the abort check) I think its cleaner if it remains inside the locked environment in the parallel case.

Reply:
Would it make sense to wrap _everything_ inside the mutex? In the non-parallel case, this would be a no-op. Since there's a lot of duplication, this could clean things up a bit (although I will definitely add a comment about the race conditions) :slightly_smiling_face: